### PR TITLE
Improve transform performance by moving tokenizer and `VideoToTensor` to GPU

### LINKED ISSUE #401: Performance bottleneck: apply_transforms running partially on CPU instead of GPU
### 🐛 Describe the bug

While testing inference performance, I noticed that `apply_transforms` adds significant latency to each inference call.
After profiling, I found that two parts of the transform pipeline are CPU-bound even when a GPU is available:

1. The **tokenizer** used by `GR00TTransform` is instantiated on the CPU.
2. The **`VideoToTensor`** transform also processes frames on the CPU.

Both operations can safely be executed on the GPU for faster preprocessing.

---

### **Steps to Reproduce**

1. Run `scripts/inference_service.py` (both server and client) as described in the README.
2. Add simple timing around the transform step in `Gr00tPolicy.apply_transforms`:

   ```python
   t = time.time()
   normalized_input = self.apply_transforms(obs_copy)
   print(f"Applied transforms in {time.time() - t:.3f}s")
   ```
3. Observe transform latency (around 50–100 ms per call on main branch).
   Note this was observed on a desktop machine with a 16GB 5080 GPU and 64GB/RAM.

Note also, I'm working on a PR to fix this issue and will put it up shortly with a link here.

### Versions

Python 3.10.18
absl-py==2.3.1
accelerate==1.2.1
aiosignal==1.4.0
albucore==0.0.17
albumentations==1.4.18
annotated-types==0.7.0
antlr4-python3-runtime==4.9.3
astunparse==1.6.3
attrs==25.4.0
av==12.3.0
blessings==1.7
cachetools==6.2.0
certifi==2025.10.5
charset-normalizer==3.4.3
click==8.3.0
cloudpickle==3.1.1
contourpy==1.3.2
cramjam==2.11.0
cycler==0.12.1
decord==0.6.0
diffusers==0.30.2
dm-tree==0.1.8
docker-pycreds==0.4.0
docstring_parser==0.17.0
einops==0.8.1
eval_type_backport==0.2.2
exceptiongroup==1.3.0
Farama-Notifications==0.0.4
fastparquet==2024.11.0
filelock==3.20.0
flash_attn==2.8.3
flatbuffers==25.9.23
fonttools==4.60.1
frozenlist==1.8.0
fsspec==2025.9.0
fvcore==0.1.5.post20221221
gast==0.6.0
gitdb==4.0.12
GitPython==3.1.45
google-auth==2.41.1
google-auth-oauthlib==1.2.2
google-pasta==0.2.0
-e git+ssh://git@github.com/bbitmaster/Isaac-GR00T.git@d03edd11ed236a0e696384da42120629e69b09ef#egg=gr00t
grpcio==1.75.1
gymnasium==1.0.0
h5py==3.12.1
hf-xet==1.1.10
huggingface-hub==0.35.3
hydra-core==1.3.2
idna==3.10
imageio==2.34.2
importlib_metadata==8.7.0
iniconfig==2.1.0
iopath==0.1.10
Jinja2==3.1.6
jsonschema==4.25.1
jsonschema-specifications==2025.9.1
keras==2.15.0
kiwisolver==1.4.9
kornia==0.7.4
kornia_rs==0.1.9
lazy_loader==0.4
libclang==18.1.1
llvmlite==0.45.1
Markdown==3.9
markdown-it-py==4.0.0
MarkupSafe==3.0.3
matplotlib==3.10.0
mdurl==0.1.2
ml-dtypes==0.2.0
mpmath==1.3.0
msgpack==1.1.2
networkx==3.4.2
numba==0.62.1
numpy==1.26.4
numpydantic==1.6.7
nvidia-cublas-cu12==12.8.3.14
nvidia-cuda-cupti-cu12==12.8.57
nvidia-cuda-nvrtc-cu12==12.8.61
nvidia-cuda-runtime-cu12==12.8.57
nvidia-cudnn-cu12==9.7.1.26
nvidia-cufft-cu12==11.3.3.41
nvidia-cufile-cu12==1.13.0.11
nvidia-curand-cu12==10.3.9.55
nvidia-cusolver-cu12==11.7.2.55
nvidia-cusparse-cu12==12.5.7.53
nvidia-cusparselt-cu12==0.6.3
nvidia-nccl-cu12==2.26.2
nvidia-nvjitlink-cu12==12.8.61
nvidia-nvtx-cu12==12.8.55
oauthlib==3.3.1
omegaconf==2.3.0
onnx==1.17.0
opencv-python==4.8.0.74
opencv-python-headless==4.11.0.86
opt_einsum==3.4.0
packaging==25.0
pandas==2.2.3
peft==0.17.0
pettingzoo==1.25.0
pillow==11.3.0
pipablepytorch3d==0.7.6
platformdirs==4.5.0
pluggy==1.6.0
portalocker==3.2.0
protobuf==3.20.3
psutil==7.1.0
pyarrow==14.0.1
pyasn1==0.6.1
pyasn1_modules==0.4.2
pydantic==2.10.6
pydantic_core==2.27.2
Pygments==2.19.2
pyparsing==3.2.5
pytest==8.4.2
python-dateutil==2.9.0.post0
pytz==2025.2
PyYAML==6.0.2
pyzmq==27.1.0
ray==2.40.0
referencing==0.36.2
regex==2025.9.18
requests==2.32.3
requests-oauthlib==2.0.0
rich==14.2.0
rpds-py==0.27.1
rsa==4.9.1
safetensors==0.6.2
scikit-image==0.25.2
scipy==1.15.3
sentry-sdk==2.41.0
setproctitle==1.3.7
shtab==1.7.2
six==1.17.0
smmap==5.0.2
sympy==1.13.3
tabulate==0.9.0
tensorboard==2.15.2
tensorboard-data-server==0.7.2
tensorflow==2.15.0
tensorflow-estimator==2.15.0
tensorflow-io-gcs-filesystem==0.37.1
termcolor==3.1.0
tianshou==0.5.1
tifffile==2025.5.10
timm==1.0.14
tokenizers==0.21.4
tomli==2.3.0
torch==2.7.1+cu128
torchaudio==2.7.1+cu128
TorchCodec==0.1.0
torchvision==0.22.1+cu128
tqdm==4.67.1
transformers==4.51.3
triton==3.3.1
typeguard==4.4.2
typing_extensions==4.12.2
tyro==0.9.17
tzdata==2025.2
urllib3==2.5.0
wandb==0.18.0
Werkzeug==3.1.3
wrapt==1.14.2
yacs==0.1.8
zipp==3.23.0

### PR DESCRIPTION
**Pull request by bbitmaster**
Fixes #401

Changes proposed in this pull request:

- Move tokenizer in `GR00TTransform` to GPU (`device="cuda"`) when available.
- Move tensors in `VideoToTensor` to GPU (`.to("cuda")`) to eliminate unnecessary CPU-GPU transfers.
- These changes reduce transform latency by up to **6.5×** without altering any outputs or functionality.

---

### Background
While profiling inference performance, I noticed that `apply_transforms` in `Gr00tPolicy` was consistently slow (50–100 ms per call).  
This was traced to two CPU-bound operations that should be GPU-accelerated:
1. The tokenizer used by `GR00TTransform` was initialized on the CPU.  
2. `VideoToTensor` performed tensor conversions on the CPU.  

Both are safe to run on GPU and significantly speed up preprocessing.

---

### Benchmark Results

| Configuration                         | Avg Time (ms) | Std. Dev. (ms) | Relative Speed |
| ------------------------------------- | ------------- | --------------- | --------------- |
| **Main branch**                       | 57.9          | 46.6            | 1.0×            |
| **Tokenizer on CUDA**                 | 39.2          | 23.0            | 1.5×            |
| **VideoToTensor `.to("cuda")`**       | 31.6          | 20.5            | 1.8×            |
| **Tokenizer + VideoToTensor on CUDA** | **8.9**       | **5.1**         | **6.5× faster** ✅ |

**Benchmark details:**  
- Each configuration was run **20 times**, averaging the total transform duration measured by inserting a `time.time()` timer around `apply_transforms` in `Gr00tPolicy`.  
- Tests were executed using the provided `scripts/inference_service.py` (server and client) following the steps in the README.  

**Environment:**  
- GPU: NVIDIA RTX 5080 (16 GB)  
- CPU: Intel Core Ultra 9 285K  
- RAM: 64 GB  
- CUDA 12.8, PyTorch 2.7.1+cu128

---

## Before submitting

- [x] I've read and followed all steps in the [Making a pull request](https://github.com/NVIDIA/Isaac-GR00T/blob/main/CONTRIBUTING.md#making-a-pull-request) section of the `CONTRIBUTING` docs.  
- [x] I've updated or added any relevant docstrings.  
- [ ] If this PR fixes a bug, I've added a test that will fail without my fix.  
      → *Not applicable — this change only improves performance, not behavior.*  
- [ ] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.  
      → *Not applicable — no new functionality added.*

### CODE REVIEW

**Comment by lcf2610** [comment]
Using this commit in npu can also gain significantly speed up, thanks.
btw, I'd like to know whether the modifications have been widely tested to determine if they cause any loss of accuracy?
